From: Daniel van Vugt <daniel.van.vugt@canonical.com>
Date: Fri, 30 Jan 2026 15:59:19 +0800
Subject: Revert "clutter/frame-clock: Remove min_render_time_allowed_us"

This reverts commit a5b62cffbd07bb12e81a5986ee4174a5a6c13e96, restoring
the distinction between min allowed and max expected render times.

Under heavy load and with triple buffering enabled,
`max_update_time_estimate_us` stays close to two frames. So the loop
was iterating slightly too many times and choosing a
`next_presentation_time_us` too far in the future causing some frames
to be skipped which were actually achievable.

Bug: https://gitlab.gnome.org/GNOME/mutter/-/issues/4581
Fixes: a5b62cffbd ("clutter/frame-clock: Remove min_render_time_allowed_us")
Origin: https://gitlab.gnome.org/GNOME/mutter/-/merge_requests/4871
---
 clutter/clutter/clutter-frame-clock.c | 9 ++++++++-
 1 file changed, 8 insertions(+), 1 deletion(-)

diff --git a/clutter/clutter/clutter-frame-clock.c b/clutter/clutter/clutter-frame-clock.c
index 88ee13b..6066d97 100644
--- a/clutter/clutter/clutter-frame-clock.c
+++ b/clutter/clutter/clutter-frame-clock.c
@@ -832,6 +832,7 @@ calculate_next_update_time_us (ClutterFrameClock *frame_clock,
   int64_t last_presentation_time_us;
   int64_t now_us;
   int64_t refresh_interval_us;
+  int64_t min_render_time_allowed_us;
   int64_t max_update_time_estimate_us;
   int64_t next_presentation_time_us;
   int64_t next_smooth_presentation_time_us = 0;
@@ -864,6 +865,11 @@ calculate_next_update_time_us (ClutterFrameClock *frame_clock,
       return;
     }
 
+  min_render_time_allowed_us = refresh_interval_us / 2;
+
+  if (min_render_time_allowed_us > max_update_time_estimate_us)
+    min_render_time_allowed_us = max_update_time_estimate_us;
+
   /*
    * The common case is that the next presentation happens 1 refresh interval
    * after the last presentation:
@@ -953,10 +959,11 @@ calculate_next_update_time_us (ClutterFrameClock *frame_clock,
       if (out_is_target_presentation_time)
         *out_is_target_presentation_time = FALSE;
       *out_next_frame_deadline_us = 0;
+      min_render_time_allowed_us = 0;
     }
   else
     {
-      while (next_presentation_time_us - max_update_time_estimate_us < now_us)
+      while (next_presentation_time_us - min_render_time_allowed_us < now_us)
         next_presentation_time_us += refresh_interval_us;
 
       next_update_time_us = next_presentation_time_us - max_update_time_estimate_us;
